File: rtvioapk/app/src/main/java/com/rtvio/mapper/net/Protocol.kt
package com.rtvio.mapper.net

import java.io.DataInputStream
import java.io.EOFException
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * The RTVIO wire format.
 *
 * Every multi-byte field is big-endian (network byte order), which is also
 * java.nio's default, so no explicit ordering call is strictly required - it is
 * stated anyway so the intent survives a future refactor.
 *
 * All three payload types are multiplexed onto one TCP stream. A reader
 * dispatches on the first byte of each packet; because a single writer thread
 * emits whole packets atomically, a frame is never interleaved with an IMU
 * batch.
 */
object Protocol {

    /**
     * v1: stream from the moment the socket connects; nothing is ever read
     * after the handshake. v2 (the desktop's rtvio.studio): connect "armed"
     * and stream only between START and STOP commands, reporting state in
     * STATUS packets. The app picks the behaviour from the version the
     * desktop greets with, so v1 receivers keep working unchanged.
     */
    const val VERSION = 2
    const val VERSION_REMOTE_CONTROL = 2

    const val HEADER_FRAME: Byte = 0xFF.toByte()
    const val HEADER_IMU: Byte = 0xFE.toByte()
    const val HEADER_GPS: Byte = 0xFD.toByte()
    const val HEADER_INTRINSICS: Byte = 0xFC.toByte()
    /** Phone -> desktop, JSON body (v2 only). */
    const val HEADER_STATUS: Byte = 0xFB.toByte()
    /** Phone -> desktop, frame layout, a viewfinder image never recorded (v2 only). */
    const val HEADER_PREVIEW: Byte = 0xFA.toByte()
    /** Desktop -> phone, JSON body (v2 only). */
    const val HEADER_COMMAND: Byte = 0xC0.toByte()
    const val HEADER_HANDSHAKE_ACK: Byte = 0xAA.toByte()

    /**
     * Phone -> desktop, one-shot bulk transfer of a locally-recorded session
     * directory (see [com.rtvio.mapper.capture.LocalSessionRecorder]). Sent on
     * its own connection, never mixed with live streaming: SESSION_BEGIN, then
     * one SESSION_FILE header + raw bytes per file, then SESSION_END.
     */
    const val HEADER_SESSION_BEGIN: Byte = 0xE0.toByte()
    const val HEADER_SESSION_FILE: Byte = 0xE1.toByte()
    const val HEADER_SESSION_END: Byte = 0xE2.toByte()

    /** Largest JSON body a u16 length field can carry. */
    const val MAX_JSON_BYTES = 65535

    /** Bytes a frame packet costs on top of the JPEG payload. */
    const val FRAME_OVERHEAD = 1 + 8 + 4 + 4 + 4      // 21
    /** Bytes an IMU batch costs on top of its samples. */
    const val IMU_OVERHEAD = 1 + 2                    // 3
    /** Bytes one IMU sample occupies. */
    const val IMU_SAMPLE_SIZE = 8 + 4 * 6             // 32
    /** Total size of a GPS packet. */
    const val GPS_PACKET_SIZE = 1 + 8 + 8 + 8 + 4 + 4 // 33
    /** Total size of the handshake ack the desktop sends on connect. */
    const val HANDSHAKE_ACK_SIZE = 1 + 4 + 1          // 6
    /**
     * Intrinsics packet: 0xFC + 4 floats (fx, fy, cx, cy) + 5 doubles (k1, k2, p1, p2, k3)
     *                  = 1 + 16 + 40 = 57 bytes, plus a null-terminated source string
     */

    /**
     * struct FramePacket { u8 0xFF; i64 timestamp_ms; i32 w; i32 h; i32 size; u8 jpeg[size]; }
     *
     * [timestampMs] is wall-clock (System.currentTimeMillis()), matching the
     * spec. Note this is a *different clock* from the IMU timestamps - see
     * [encodeImuBatch].
     */
    fun encodeFrame(
        timestampMs: Long,
        width: Int,
        height: Int,
        jpeg: ByteArray,
        jpegLength: Int = jpeg.size,
        header: Byte = HEADER_FRAME
    ): ByteArray {
        val buf = ByteBuffer.allocate(FRAME_OVERHEAD + jpegLength).order(ByteOrder.BIG_ENDIAN)
        buf.put(header)
        buf.putLong(timestampMs)
        buf.putInt(width)
        buf.putInt(height)
        buf.putInt(jpegLength)
        buf.put(jpeg, 0, jpegLength)
        return buf.array()
    }

    /**
     * struct IMUBatchPacket { u8 0xFE; i16 count; { i64 t_ns; f32 a[3]; f32 g[3]; }[count]; }
     *
     * Sample timestamps are SensorEvent.timestamp - nanoseconds since boot, on
     * the monotonic clock, NOT epoch. That is what the spec asks for and it is
     * the right choice (it does not jump when NTP corrects the wall clock), but
     * it means the desktop must recover the offset between the two clocks
     * before it can associate an IMU sample with a frame. See README
     * "Clock domains" for how to do that from the stream alone.
     *
     * @param count how many entries of [samples] are populated; the array may
     *   be a reusable over-allocated scratch buffer.
     */
    fun encodeImuBatch(samples: List<ImuSample>, count: Int = samples.size): ByteArray {
        require(count <= Short.MAX_VALUE) { "IMU batch of $count exceeds the i16 count field" }
        val buf = ByteBuffer.allocate(IMU_OVERHEAD + count * IMU_SAMPLE_SIZE)
            .order(ByteOrder.BIG_ENDIAN)
        buf.put(HEADER_IMU)
        buf.putShort(count.toShort())
        for (i in 0 until count) {
            val s = samples[i]
            buf.putLong(s.timestampNs)
            buf.putFloat(s.ax); buf.putFloat(s.ay); buf.putFloat(s.az)
            buf.putFloat(s.gx); buf.putFloat(s.gy); buf.putFloat(s.gz)
        }
        return buf.array()
    }

    /**
     * struct GPSPacket { u8 0xFD; i64 t_ms; f64 lat; f64 lon; f32 alt; f32 acc; }
     *
     * [timestampMs] is Location.getTime(), i.e. epoch milliseconds from the GNSS
     * fix itself - the same clock domain as the frame timestamps.
     */
    fun encodeGps(
        timestampMs: Long,
        latitude: Double,
        longitude: Double,
        altitudeM: Float,
        accuracyM: Float
    ): ByteArray {
        val buf = ByteBuffer.allocate(GPS_PACKET_SIZE).order(ByteOrder.BIG_ENDIAN)
        buf.put(HEADER_GPS)
        buf.putLong(timestampMs)
        buf.putDouble(latitude)
        buf.putDouble(longitude)
        buf.putFloat(altitudeM)
        buf.putFloat(accuracyM)
        return buf.array()
    }

    /**
     * struct IntrinsicsPacket {
     *   u8 0xFC;
     *   f32 fx_pix, fy_pix, cx_pix, cy_pix;
     *   f64 k1, k2, p1, p2, k3;
     *   u16 source_len;
     *   u8 source[source_len];  // null-terminated string describing the source
     * }
     *
     * Sent once at session start, right after the phone connects. Contains the
     * camera intrinsic matrix K and distortion coefficients. Always preferred over
     * any cached calibration or hardcoded defaults on the desktop.
     */
    fun encodeIntrinsics(
        fxPix: Float,
        fyPix: Float,
        cxPix: Float,
        cyPix: Float,
        k1: Double = 0.0,
        k2: Double = 0.0,
        p1: Double = 0.0,
        p2: Double = 0.0,
        k3: Double = 0.0,
        source: String = ""
    ): ByteArray {
        val sourceBytes = source.toByteArray(Charsets.UTF_8)
        val buf = ByteBuffer.allocate(1 + 16 + 40 + 2 + sourceBytes.size).order(ByteOrder.BIG_ENDIAN)
        buf.put(HEADER_INTRINSICS)
        buf.putFloat(fxPix)
        buf.putFloat(fyPix)
        buf.putFloat(cxPix)
        buf.putFloat(cyPix)
        buf.putDouble(k1)
        buf.putDouble(k2)
        buf.putDouble(p1)
        buf.putDouble(p2)
        buf.putDouble(k3)
        buf.putShort(sourceBytes.size.toShort())
        buf.put(sourceBytes)
        return buf.array()
    }

    /**
     * struct StatusPacket { u8 0xFB; u16 len; u8 json[len]; }
     *
     * [json] is built by the caller (org.json at runtime); this only frames
     * it, which keeps the unit-testable part free of the android.jar stubs.
     */
    fun encodeStatus(json: String): ByteArray {
        val body = json.toByteArray(Charsets.UTF_8)
        require(body.size <= MAX_JSON_BYTES) { "status body of ${body.size} bytes exceeds u16" }
        return ByteBuffer.allocate(3 + body.size).order(ByteOrder.BIG_ENDIAN)
            .put(HEADER_STATUS)
            .putShort(body.size.toShort())
            .put(body)
            .array()
    }

    /**
     * struct SessionBeginPacket { u8 0xE0; u16 idLen; u8 id[idLen]; i32 fileCount; i64 totalBytes; }
     *
     * Opens a session transfer. [totalBytes] is the sum of every file about to
     * follow, purely so the receiver can print progress; nothing on the wire
     * depends on it being exact.
     */
    fun encodeSessionBegin(sessionId: String, fileCount: Int, totalBytes: Long): ByteArray {
        val idBytes = sessionId.toByteArray(Charsets.UTF_8)
        return ByteBuffer.allocate(1 + 2 + idBytes.size + 4 + 8).order(ByteOrder.BIG_ENDIAN)
            .put(HEADER_SESSION_BEGIN)
            .putShort(idBytes.size.toShort())
            .put(idBytes)
            .putInt(fileCount)
            .putLong(totalBytes)
            .array()
    }

    /**
     * struct SessionFilePacket { u8 0xE1; u16 pathLen; u8 path[pathLen]; i64 size; }
     *
     * [relativePath] uses '/' separators regardless of host OS, and is
     * immediately followed on the wire by exactly [fileSize] raw bytes - no
     * further framing, since the size is already known.
     */
    fun encodeSessionFileHeader(relativePath: String, fileSize: Long): ByteArray {
        val pathBytes = relativePath.toByteArray(Charsets.UTF_8)
        return ByteBuffer.allocate(1 + 2 + pathBytes.size + 8).order(ByteOrder.BIG_ENDIAN)
            .put(HEADER_SESSION_FILE)
            .putShort(pathBytes.size.toShort())
            .put(pathBytes)
            .putLong(fileSize)
            .array()
    }

    /** struct SessionEndPacket { u8 0xE2; } - every file has been sent. */
    fun encodeSessionEnd(): ByteArray = byteArrayOf(HEADER_SESSION_END)

    /**
     * Reads one desktop -> phone packet: u8 0xC0; u16 len; u8 json[len].
     * Blocks until a full packet arrives. Returns the JSON text; throws
     * [java.io.IOException] on a closed stream or an unexpected header (the
     * desktop sends nothing else after the handshake, so anything else means
     * the stream is out of sync).
     */
    fun readCommand(input: DataInputStream): String {
        val header = input.readByte()
        if (header != HEADER_COMMAND) {
            throw java.io.IOException("unexpected packet 0x%02X from desktop".format(header.toInt() and 0xFF))
        }
        val len = input.readUnsignedShort()
        val body = ByteArray(len)
        input.readFully(body)
        return String(body, Charsets.UTF_8)
    }

    /** Result of reading the desktop's greeting. */
    data class HandshakeAck(val protocolVersion: Int, val status: Int) {
        val ok: Boolean get() = status == 0
    }

    /**
     * Reads the 6-byte ack the desktop is specified to send on connect.
     *
     * Returns null if the peer sent something that is not an ack, or closed
     * without sending one. A null is deliberately not fatal: a receiver that
     * simply starts reading without greeting us is still a usable receiver, so
     * the caller reports "no handshake" rather than refusing to stream.
     */
    fun readHandshakeAck(input: DataInputStream): HandshakeAck? = try {
        val header = input.readByte()
        if (header != HEADER_HANDSHAKE_ACK) {
            null
        } else {
            // uint32 read into a Long so 0x80000000+ does not come back negative.
            val version = input.readInt().toLong() and 0xFFFFFFFFL
            val status = input.readByte().toInt() and 0xFF
            HandshakeAck(version.toInt(), status)
        }
    } catch (e: EOFException) {
        null
    }
}

/** One time-aligned inertial sample: accelerometer m/s^2, gyroscope rad/s. */
data class ImuSample(
    val timestampNs: Long,
    val ax: Float, val ay: Float, val az: Float,
    val gx: Float, val gy: Float, val gz: Float
)

File: rtvio/src/rtvio/stream/protocol.py
"""
RTVIO wire protocol: bytes <-> typed packets.

Canonical definition is rtvioapk/.../net/Protocol.kt; this is the Python
side of the same contract, and the struct formats below are byte-for-byte
what rtvioapk/tools/mock_receiver.py already validates end-to-end against
the app. All fields are big-endian.

Kept deliberately free of sockets and of session state so the same parser
serves the live socket, the replay driver and the unit tests. The only
socket-aware function is `recv_exact`, and it is here because getting it
wrong is the single most common bug when writing a receiver:
**sock.recv(n) may return fewer than n bytes.** A 200 KB JPEG essentially
always arrives split across segments; reading a length field and calling
recv once works on localhost and fails over WiFi.
"""
import json
import struct
from typing import NamedTuple

HEADER_FRAME = 0xFF
HEADER_IMU = 0xFE
HEADER_GPS = 0xFD
HEADER_INTRINSICS = 0xFC
HEADER_STATUS = 0xFB           # phone -> desktop, JSON (protocol v2 only)
HEADER_PREVIEW = 0xFA          # phone -> desktop, same layout as FRAME; a
                               # low-rate viewfinder image sent while armed,
                               # never recorded (protocol v2 only)
HEADER_COMMAND = 0xC0          # desktop -> phone, JSON (protocol v2 only)
HEADER_HANDSHAKE_ACK = 0xAA
# phone -> desktop, one-shot bulk copy of a RECORD LOCALLY session (the app's
# Saved sessions -> Transfer action), its own dedicated connection, never
# mixed with live streaming. Canonical definition is Protocol.kt; wire format
# matches rtvioapk/tools/mock_receiver.py's receive_session_transfer exactly:
# SESSION_BEGIN (pstring id, i32 file_count, i64 total_bytes), then
# file_count x SESSION_FILE (pstring rel_path, i64 size, raw bytes), then
# SESSION_END.
HEADER_SESSION_BEGIN = 0xE0
HEADER_SESSION_FILE = 0xE1
HEADER_SESSION_END = 0xE2

# v1: the phone streams as soon as it connects and never reads another byte
# after the handshake. v2 (rtvio.studio): the phone connects "armed", streams
# only between a START and a STOP command, and reports its state in STATUS
# packets. The app decides which behaviour to use from the version in the
# handshake, so a v1 receiver (live_pipeline.py, mock_receiver.py) keeps
# working with the new app unchanged.
PROTOCOL_VERSION = 1
STUDIO_PROTOCOL_VERSION = 2
STATUS_OK = 0

HANDSHAKE_FMT = ">BIB"          # header, version, status
FRAME_HEADER_FMT = ">qiii"      # timestamp_ms, width, height, jpeg_size
IMU_COUNT_FMT = ">h"            # sample count
IMU_SAMPLE_FMT = ">qffffff"     # t_ns, ax, ay, az, gx, gy, gz
GPS_FMT = ">qddff"              # timestamp_ms, lat, lon, alt_m, accuracy_m

FRAME_HEADER_BYTES = struct.calcsize(FRAME_HEADER_FMT)
IMU_SAMPLE_BYTES = struct.calcsize(IMU_SAMPLE_FMT)
GPS_BYTES = struct.calcsize(GPS_FMT)

# Sanity ceilings so a desynchronised stream cannot make us allocate wildly.
MAX_JPEG_BYTES = 32 * 1024 * 1024
MAX_IMU_BATCH = 4096


class StreamClosed(Exception):
    """The peer hung up, or the stream desynchronised past recovery."""


class FramePacket(NamedTuple):
    """A JPEG still. `timestamp_ms` is WALL CLOCK (System.currentTimeMillis)."""
    timestamp_ms: int
    width: int
    height: int
    jpeg: bytes


class ImuSample(NamedTuple):
    """One accelerometer sample with the gyro already interpolated onto its
    timestamp by the app (see SensorDataCollector's class comment) - so the
    two triples are genuinely simultaneous and need no re-alignment here.

    `t_ns` is MONOTONIC SINCE BOOT (SensorEvent.timestamp), a different
    clock domain from every other packet. See clock.py.

    accel is m/s^2 and INCLUDES GRAVITY (TYPE_ACCELEROMETER, not
    TYPE_LINEAR_ACCELERATION), which is what so3.level_and_align_attitude
    requires. gyro is rad/s.
    """
    t_ns: int
    accel: tuple
    gyro: tuple


class GpsPacket(NamedTuple):
    """`timestamp_ms` is WALL CLOCK (Location.getTime). `altitude_m` is
    documented as metres above the WGS84 ellipsoid, but vendor
    implementations vary and some return orthometric height - a 20-100 m
    systematic Z offset that is very hard to spot downstream.

    `accuracy_m` is a metre-scale 1-sigma horizontal estimate, or -1 when
    unknown. live_pipeline.py's on_gps re-anchors the pose directly on each
    fix rather than fusing it, so this value isn't consumed as a filter
    covariance any more - kept on the wire (and in AltitudeSanity/reports)
    as a real per-fix quality signal.
    """
    timestamp_ms: int
    lat_deg: float
    lon_deg: float
    altitude_m: float
    accuracy_m: float

    @property
    def accuracy_known(self):
        return self.accuracy_m >= 0.0


class IntrinsicsPacket(NamedTuple):
    """Camera intrinsic matrix K and distortion coefficients from the phone.

    Sent once at session start. The phone has measured these from Camera2 API
    or computed from focal length and sensor size; either way, they are the
    ground truth for this device's camera at this resolution.

    K = [[fx_pix, 0, cx_pix], [0, fy_pix, cy_pix], [0, 0, 1]]
    distortion = [k1, k2, p1, p2, k3]
    """
    fx_pix: float
    fy_pix: float
    cx_pix: float
    cy_pix: float
    k1: float
    k2: float
    p1: float
    p2: float
    k3: float
    source: str  # Description of how these were obtained


def recv_exact(sock, count):
    """Read exactly `count` bytes from `sock`, or raise StreamClosed."""
    chunks = []
    remaining = count
    while remaining:
        chunk = sock.recv(remaining)
        if not chunk:
            raise StreamClosed(f"peer closed with {remaining} of {count} bytes outstanding")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def read_pstring(sock):
    """u16 length + utf-8 bytes, the framing every string field on the wire
    uses (session id, file relative paths)."""
    (n,) = struct.unpack(">H", recv_exact(sock, 2))
    return recv_exact(sock, n).decode("utf-8") if n else ""


def encode_handshake(version=PROTOCOL_VERSION):
    """Desktop -> phone greeting. A v1 app reads exactly these 6 bytes and
    never expects another inbound byte. Only send version=2 from a receiver
    that also speaks COMMAND/STATUS (rtvio.studio): a v2-aware app treats
    that version as "wait for a START command" instead of streaming."""
    return struct.pack(HANDSHAKE_FMT, HEADER_HANDSHAKE_ACK, version, STATUS_OK)


# JSON-bodied control packets (v2). u8 header, u16 length, UTF-8 JSON object.
# JSON rather than fixed structs because these carry a growing set of
# optional fields (capture settings, stats) and are sent a few times a
# second at most - the byte-exact fixed layouts above are for the hot path.
JSON_LEN_FMT = ">H"
MAX_JSON_BYTES = 65535


def _encode_json(header, obj):
    body = json.dumps(obj, separators=(",", ":")).encode("utf-8")
    if len(body) > MAX_JSON_BYTES:
        raise ValueError("control packet body is %d bytes, max %d" % (len(body), MAX_JSON_BYTES))
    return bytes([header]) + struct.pack(JSON_LEN_FMT, len(body)) + body


def _read_json(sock):
    (n,) = struct.unpack(JSON_LEN_FMT, recv_exact(sock, 2))
    body = recv_exact(sock, n) if n else b"{}"
    try:
        obj = json.loads(body.decode("utf-8"))
    except (UnicodeDecodeError, ValueError) as e:
        raise StreamClosed("malformed JSON control packet: %s" % e)
    if not isinstance(obj, dict):
        raise StreamClosed("control packet body is not a JSON object")
    return obj


def encode_command(obj):
    """Desktop -> phone. obj["cmd"] is one of "start", "stop", "ping"."""
    return _encode_json(HEADER_COMMAND, obj)


def encode_status(obj):
    """Phone -> desktop. Used by the phone simulator and the tests; the real
    encoder is Protocol.encodeStatus in the app."""
    return _encode_json(HEADER_STATUS, obj)


def read_command(sock):
    """Body of a COMMAND packet whose header byte was already consumed."""
    return _read_json(sock)


def read_status(sock):
    """Body of a STATUS packet whose header byte was already consumed."""
    return _read_json(sock)


# ------------------------------------------------------------------ decode --

def read_frame(sock):
    timestamp_ms, width, height, jpeg_size = struct.unpack(
        FRAME_HEADER_FMT, recv_exact(sock, FRAME_HEADER_BYTES))
    if not 0 < jpeg_size <= MAX_JPEG_BYTES:
        raise StreamClosed(f"implausible jpeg_size {jpeg_size}; stream desynchronised")
    if width <= 0 or height <= 0:
        raise StreamClosed(f"implausible frame size {width}x{height}; stream desynchronised")
    return FramePacket(timestamp_ms, width, height, recv_exact(sock, jpeg_size))


def read_imu(sock):
    (count,) = struct.unpack(IMU_COUNT_FMT, recv_exact(sock, 2))
    if not 0 <= count <= MAX_IMU_BATCH:
        raise StreamClosed(f"implausible IMU sample count {count}; stream desynchronised")
    payload = recv_exact(sock, count * IMU_SAMPLE_BYTES)
    out = []
    for i in range(count):
        t_ns, ax, ay, az, gx, gy, gz = struct.unpack_from(
            IMU_SAMPLE_FMT, payload, i * IMU_SAMPLE_BYTES)
        out.append(ImuSample(t_ns, (ax, ay, az), (gx, gy, gz)))
    return out


def read_gps(sock):
    return GpsPacket(*struct.unpack(GPS_FMT, recv_exact(sock, GPS_BYTES)))


def read_intrinsics(sock):
    """Read 0xFC intrinsics packet: 4 floats (K) + 5 doubles (distortion) + source string."""
    # 4 floats (fx, fy, cx, cy)
    k_floats = struct.unpack(">ffff", recv_exact(sock, 16))
    # 5 doubles (k1, k2, p1, p2, k3)
    distortion = struct.unpack(">ddddd", recv_exact(sock, 40))
    # source string: u16 length + bytes
    source_len_bytes = recv_exact(sock, 2)
    source_len = struct.unpack(">H", source_len_bytes)[0]
    source_bytes = recv_exact(sock, source_len) if source_len > 0 else b""
    source = source_bytes.decode("utf-8")

    return IntrinsicsPacket(
        fx_pix=k_floats[0],
        fy_pix=k_floats[1],
        cx_pix=k_floats[2],
        cy_pix=k_floats[3],
        k1=distortion[0],
        k2=distortion[1],
        p1=distortion[2],
        p2=distortion[3],
        k3=distortion[4],
        source=source
    )


def read_session_begin(sock):
    """Body of a SESSION_BEGIN packet whose header byte was already
    consumed: (session_id, file_count, total_bytes)."""
    session_id = read_pstring(sock)
    file_count, total_bytes = struct.unpack(">iq", recv_exact(sock, 12))
    return session_id, file_count, total_bytes


def read_session_file_header(sock):
    """Body of a SESSION_FILE packet whose header byte was already
    consumed: (relative_path, file_size) - the raw file bytes follow and
    are the caller's to read (recv_exact would buffer a large video/frame
    dump entirely in memory; the caller streams it to disk instead)."""
    rel_path = read_pstring(sock)
    (file_size,) = struct.unpack(">q", recv_exact(sock, 8))
    return rel_path, file_size


def read_packet(sock):
    """Read one packet. Returns (header_byte, payload) where payload is a
    FramePacket, a list[ImuSample], a GpsPacket, or an IntrinsicsPacket."""
    header = recv_exact(sock, 1)[0]
    if header == HEADER_FRAME:
        return header, read_frame(sock)
    if header == HEADER_IMU:
        return header, read_imu(sock)
    if header == HEADER_GPS:
        return header, read_gps(sock)
    if header == HEADER_INTRINSICS:
        return header, read_intrinsics(sock)
    if header == HEADER_STATUS:
        return header, read_status(sock)
    raise StreamClosed(f"unknown packet header 0x{header:02X}")


# ------------------------------------------------------------------ encode --
# Used by tools/replay_dataset_as_phone.py to stand in for the handset, and
# by the round-trip tests. Keeping encode and decode in one file is what
# makes "the test double speaks the same protocol" checkable rather than
# aspirational.

def encode_frame(pkt: FramePacket):
    return (bytes([HEADER_FRAME])
            + struct.pack(FRAME_HEADER_FMT, pkt.timestamp_ms, pkt.width,
                          pkt.height, len(pkt.jpeg))
            + pkt.jpeg)


def encode_imu_batch(samples):
    out = [bytes([HEADER_IMU]), struct.pack(IMU_COUNT_FMT, len(samples))]
    for s in samples:
        out.append(struct.pack(IMU_SAMPLE_FMT, s.t_ns, *s.accel, *s.gyro))
    return b"".join(out)


def encode_gps(pkt: GpsPacket):
    return bytes([HEADER_GPS]) + struct.pack(
        GPS_FMT, pkt.timestamp_ms, pkt.lat_deg, pkt.lon_deg,
        pkt.altitude_m, pkt.accuracy_m)
